apps/labs/posts/scaling-asyncio-on-free-threaded-python.md

```diff
@@ -0,0 +1,169 @@
+---
+title: 'Scaling asyncio on Free-Threaded Python'
+authors: [kumar-aditya]
+published: August 28, 2025
+description: 'A recap on the work done in Python 3.14 to enable asyncio to scale on the free-threaded build of CPython.'
+category: [Community]
+featuredImage:
+  src: /posts/scaling-asyncio-on-free-threaded-python/logo.png
+  alt: 'A logo for asyncio.'
+hero:
+  imageSrc: /posts/scaling-asyncio-on-free-threaded-python/logo.png
+  imageAlt: 'A logo for asyncio.'
+---
+
+## Introduction
+
+`asyncio` is a Python standard library for writing high performance concurrent
+code using the async/await syntax and provides high level APIs for creating and
+managing event loops, coroutines, tasks, and performing asynchronous I/O
+operations. It is used as a foundation for Python asynchronous frameworks that
+provide high-performance network and web-servers, database connection
+libraries, distributed task queues, etc. Multiple libraries and frameworks, such
+as `FastAPI` and `aiohttp`, are built on top of `asyncio`.
+
+In this blog post, we will explore the changes I made in the upcoming Python
+3.14 release to enable `asyncio` to scale on the free-threaded build of CPython.
+
+## The GIL and asyncio: A brief recap
+
+Before diving into the details of scaling `asyncio` on the free-threaded build
+of CPython, it's important to understand what is Global Interpreter Lock (GIL)
+and how it is a significant limitation for `asyncio` in the first place.
+
+The Global Interpreter Lock (GIL) is a global mutex that protects access to
+Python objects, preventing multiple threads from executing Python bytecodes at
+once. This means that even though you can have multiple threads in a Python
+program, only one thread can execute Python code at a time.
+
+`asyncio` uses an event loop as a scheduler to enable highly efficient I/O-bound
+concurrency by switching between tasks during non-blocking I/O operations. The
+event loop leverages platform-specific support for asynchronous I/O—such as
+`epoll` on Linux, `kqueue` on macOS, and `IOCP` on Windows to perform these
+operations efficiently. Since only one event loop can run per thread, CPU-bound
+tasks, which would otherwise block the event loop, are typically offloaded to
+separate threads. However, the GIL limits true parallel execution of Python code
+across threads. Hence, even when tasks are offloaded, they still compete for the
+GIL for execution and limits parallelism.
```

Comment:
there's a grammar issue in this sentence. You could just make it read "and this limits parallelism" to fix that, but I edited it a little bit and added some more elaboration. Feel free to take it or leave it.

```suggestion
GIL for execution. This lock contention limits parallelism and can tank performance for any CPU-bound workload.
```